### DISCUSSION TITLE
[question] Traffic updates

### DISCUSSION THREAD

**Discussion by vl4deee11** [Project-OSRM/osrm-backend#6823] [Q&A]
Good afternoon, first of all I will say thank you for such a wonderful thing as OSRM. Actually, the question itself, I know that it is possible to update traffic using `--segment-speed-file updates.csv`, and this is really cool! 
But can you tell me if there are paid / free solutions that allow one way or another to get the current speed from node A to node B, or coordinate or whatever else ? 
Thus , my biggest problem in building my solution is where can I get such data at all ?

**Comment by nilsnolde** [comment]
You’re looking for live traffic endpoints. I think there’s quite a few providers. I usually use TomTom‘s: https://developer.tomtom.com/intermediate-traffic-service/documentation/service/tomtom-traffic-flow-intermediate-service-protobuf. HERE has a similar product. It’s fairly expensive (though totally worth the money if you think about the their processing), usually TomTom doesn’t pick up the phone unless you spend min $10k per year.

And that’s only half of the story. Next you’ll need convert whatever you’ll get from any live traffic endpoint to what OSRM understands. Even if you have a OSRM base graph built from TomTom data, it’s a complex flow to map-match the live traffic entries to the graph to get the right IDs and match the right speeds. The format of choice is openlr, which needs some understanding of its spec.

If you’re set on doing this, it’ll be hell of a ride😄 We’ve done this in the past for some clients, so if you need commercial support, let me know on enquiry@gis-ops.com.
